pufferlib/ocean/tetris/tetris.h

```diff
@@ -0,0 +1,533 @@
+        return -1;
+    }
+    for (int r = TETROMINOES_FILLS_ROW[tetromino_idx][target_rotation]; r < env->n_cols; r++) {
+        if (env->row_is_free[r]){
+            landing_row = r;
+        }
+        else {
+            break;
+        }
+    }
+    for (int test_row = 0; test_row < env->n_rows; test_row++){
+        if (can_place_tetromino(env, tetromino_idx, target_col, target_rotation, test_row)){
+            landing_row = test_row;
+        } 
+        else {
+            break;
+        }
+    }
+    return landing_row;
+}
+
+bool is_full_row(Tetris* env, int row){
+    for (int c = 0; c < env->n_cols; c++){
+        if (env->grid[row * env->n_cols + c] == 0){
+            return false;
+        } 
+    }
+    return true;
+}
+
+void clear_row(Tetris* env, int row){
+    for (int r = row; r > 0; r--){
+        for (int c = 0; c < env->n_cols; c++){
+            env->grid[r * env->n_cols + c] = env->grid[ (r-1) * env->n_cols + c];
+        }
+    }
+    for (int c = 0; c < env->n_cols; c++){
+        env->grid[c] = 0; 
+    }
+}
+
+int place_tetromino(Tetris* env, int tetromino_idx, int col, int rotation, int landing_row) {
+    for (int r = 0; r < SIZE; r++) {
+        for (int c = 0; c < SIZE; c++) {
+            if (TETROMINOES[tetromino_idx][rotation][r][c] == 1) {
+                int grid_row = landing_row + r;
+                int grid_col = col + c;
+
+                if (grid_row < env->n_rows && grid_col >= 0 && grid_col < env->n_cols) {
+                    env->grid[grid_row * env->n_cols + grid_col] = tetromino_idx + 1;
+                    env->row_is_free[grid_row] = 0;
+                }
+            }
+        }
+    }
+    int lines_deleted = 0;
+    int row_to_check = landing_row + TETROMINOES_FILLS_ROW[tetromino_idx][rotation] - 1;
+    for (int r = 0; r < TETROMINOES_FILLS_ROW[tetromino_idx][rotation]; r++){
+        if (is_full_row(env, row_to_check)){
+            clear_row(env, row_to_check);
+            lines_deleted+=1;
+        }
+        else{
+            row_to_check-=1;
+        }
+    }
+    return lines_deleted;
+}
+
+void c_reset(Tetris* env) {
+    env->score = 0;
+    env->ep_return = 0;
+    env->step = 0;
+    env->lines_deleted = 0;
+    env->combos = 0;
+    restore_grid(env);
+    initialize_deck(env);
+    compute_action_mask(env);
+    compute_observations(env);
+}
+
+bool is_game_done(Tetris* env){
+    if (env->row_is_free[SIZE]) {
+        return false;
+    }
+    for (int action = 0; action < NUM_ROTATIONS * env->n_cols; action++) {
+        if (env->action_mask[action]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+void c_step(Tetris* env) {
+    env->terminals[0] = 0;
+    env->rewards[0] = 0.0;
+    env->step += 1;
+    int action = env->actions[0];
+    int col = action%env->n_cols;
+    int rotation = action/env->n_cols;
+    int lines_deleted = 0;
+    int landing_row;
+
+    if (env->action_mask[action]) {
```

Comment:
Waste to compute the whole atn mask if you don't need it. We don't have train-level action mask supporting right now. It's very effective in specific problems, but not generally viable because computing action mask a-priori can be slow or intractable